### ISSUE TITLE
Have jupyterhub pass options to jupyter

### ISSUE THREAD

**Issue by davidedelvento** [jupyterhub/jupyterhub#1113]
This might be a documentation issue, but I thought submitting it anyway, since even so it's still an issue.

I'm using jupyterhub `v0.7.2` for python2, 3 and R notebooks and it's generally working great, so thanks for making such a good project available.

I need my jupyter notebook to use ipywidget, but I'm stuck with the [Widget Javascript not detected](https://github.com/jupyter-widgets/ipywidgets/issues/541) issue.

The way to solve that problem is to start jupyter with the `jupyter nbextension enable --py --sys-prefix widgetsnbextension` command, however I have no control on how jupyter is started: jupyterhub does.

I've searched the documentation and the github issues without finding anything relevant. I suppose it's something that I'd write in my `jupyterhub_config.py` however I can't use introspection to see what I can do there (those `traitlets.config.loader.Config` are so annoying in this regard) nor I can find any comprehensive documentation about it (only a few examples, none of which seems applicable)

Hence the question: is it possible to have jupyterhub pass options to jupyter? If so how?

**Comment by willingc** [comment]
Hi @davidedelvento,

I suspect that there is a configuration mismatch with the Python version used by the language kernel and the hub for widgets. 

I'm going to pass this to @minrk for help on debugging your config.

**Comment by davidedelvento** [comment]
Thanks @willingc for the message. However I'm not clear what you mean, so let me rephrase my question. Is there a way (e.g. config file) to use jupyter command line options when running jupyterhub? Just using the jupyter command line options as command line options for jupyterhub does not work.

Just to be clear, I'm running the following (no virtualenv, no conda, no other versions of python installed):

```
$ python3 --version                                                                                                                                                                                                       
Python 3.4.3

$ pip3 show jupyterhub                                                                                                                                                                                                    
Name: jupyterhub
Version: 0.7.2
Summary: JupyterHub: A multi-user server for Jupyter notebooks
Home-page: http://jupyter.org
Author: Jupyter Development Team
Author-email: jupyter@googlegroups.com
License: BSD
Location: /usr/local/lib/python3.4/dist-packages
Requires: traitlets, jinja2, alembic, pamela, requests, tornado, sqlalchemy
                                                                                                    
$ pip3 show ipykernel
Name: ipykernel
Version: 4.5.2
Summary: IPython Kernel for Jupyter
Home-page: http://ipython.org
Author: IPython Development Team
Author-email: ipython-dev@scipy.org
License: BSD
Location: /usr/local/lib/python3.4/dist-packages
Requires: tornado, traitlets, ipython, jupyter-client

$ pip3 show ipywidgets
Name: ipywidgets
Version: 6.0.0
Summary: IPython HTML widgets for Jupyter
Home-page: http://ipython.org
Author: IPython Development Team
Author-email: ipython-dev@scipy.org
License: BSD
Location: /usr/local/lib/python3.4/dist-packages
Requires: ipykernel, nbformat, ipython, widgetsnbextension, traitlets
```

I do have python2 installed as well, however, I don't have jupyterhub installed for it, and the kernel I'm using for the widgets is a python3. For completeness I'm reporting what I have:

```
$ python2 --version                                                                                                                                                                                                       
Python 2.7.6

$ pip2 show jupyterhub                                                                                             

$ pip2 show ipykernel                                                                                                                                                                                                     
Name: ipykernel
Version: 4.5.2
Summary: IPython Kernel for Jupyter
Home-page: http://ipython.org
Author: IPython Development Team
Author-email: ipython-dev@scipy.org
License: BSD
Location: /usr/local/lib/python2.7/dist-packages
Requires: ipython, traitlets, tornado, jupyter-client

$ pip2 show ipywidgets                              
```

**Comment by betatim** [comment]
I think the general answer is "it depends on your spawner", as that is the bit of code responsible for starting the singleuser notebook server. Take a look at https://github.com/jupyterhub/jupyterhub/blob/master/jupyterhub/spawner.py#L230 which should work for the basic spawner (and probably others as well).

**Comment by willingc** [comment]
Sorry that I was not clear earlier. Let me try to clarify your questions to make sure that I understand them correctly. Initially, let's put the widgets aside. 

- Can JupyterHub configure, from the command line `jupyterhub`, configuration options for Jupyter notebook servers that are spawned by JupyterHub?

The spawner is in charge of creating the single user Jupyter notebook servers. Which spawner are you using?

The example [demospawner](https://github.com/jupyterhub/jupyterhub/blob/master/examples/spawn-form/jupyterhub_config.py) shows how a form could be used to set additional CLI arguments [Documentation on spawner options form](https://jupyterhub.readthedocs.io/en/latest/spawners.html#spawner-options-form)

**Comment by davidedelvento** [comment]
Thanks @willingc and @betatim this helps me understand how things works. This is progress. However it's on one hand too much and on the other too little. Too much, because I can't have (and don't want) my users specifying their custom options, I just want to use the ones I specify. Too little because I tried to reverse engineer what I needed to do, but everything failed, probably because I'm still not running jupyter, but jupyterhub-singleuser (see below).

I'm using the default spawner which is `LocalProcessSpawner` (I may later need to switch to `SudoSpawner`, but to keep things simpler I'm using the former and running as root for now).

Looking at the code that @betatim pointed out, and how it is used in the form that @willingc suggested, I tried to add to my `jupyter_config.py` various options such as

```
#c.LocalProcessSpawner.args.extend(['nbextension enable'])
#c.LocalProcessSpawner.args.extend(['nbextension', 'enable'])
#c.LocalProcessSpawner.args.extend(['nbextension', 'enable', '--sys-prefix', 'widgetsnbextension'])
```

I see from the logs that the arguments I'm setting here are correctly used to (try to) spawn a `jupyterhub-singleuser` instance, however everything fails as follows. The first two failed with errors like

```
[C 2017-05-01 12:55:58.843 ddvento.ucar.edu notebookapp:882] No such file or directory: /home/davide/nbextension
```

the last one failed with

```
[C 2017-05-01 12:57:37.595 SingleUserNotebookApp application:90] Bad config encountered during initialization:
[C 2017-05-01 12:57:37.595 SingleUserNotebookApp application:91] Unrecognized flag: '--sys-prefix'
```

So my suspicion at this point is that spawining `jupyterhub-singleuser` is not the same thing, and actually incompatible with manually invoking jupyter. Do you know if that's the case?

**Comment by davidedelvento** [comment]
Trying to dig deeper into this, it seems that many people have this problem, see for example https://github.com/ipython-contrib/jupyter_contrib_nbextensions/issues/527 https://github.com/ipython-contrib/jupyter_contrib_nbextensions/issues/617 https://github.com/ipython-contrib/jupyter_contrib_nbextensions/pull/626 https://github.com/Jupyter-contrib/jupyter_nbextensions_configurator/issues/11 https://github.com/jupyter-widgets/ipywidgets/issues/1146 https://github.com/ipython-contrib/jupyter_contrib_nbextensions/issues/941 and many more. Trying to distill all that information, the solution was already at my fingertips, namely running:

```
$ sudo jupyter nbextension install --py widgetsnbextension --sys-prefix
$ sudo jupyter nbextension enable --py widgetsnbextension --sys-prefix 
```

The thing that was confusing me very was that these commands are **not** starting jupyter with those options! Those commands were simply copying files somewhere and setting config options somewhere that then it gets picked up. So, for me the problem is solved. I still think it's poorly documented (however it may not be a jupyter thing), and therefore I'm not closing the issue, but feel free to close it yourself if it's not something this team needs to do.

Thanks again!

**Comment by willingc** [comment]
@davidedelvento Glad you solved your question.

I'm a documentation geek so I would be interested in (if you can remember back) where you first got stuck with installing and enabling ipywidgets (and its associated widgetsnbextension) with JupyterHub. That's where I would like to add documentation.

Just looking at your experience, were there any key tips or documentation that you feel would be helpful for future users configuring this. I've labeled this as a Documentation issue. (PRs are welcome too.)

Thanks!

**Comment by davidedelvento** [comment]
@willingc Thanks for asking. I don't have a good solution, but hopefully some hints that can help. I **think** I started with http://ipywidgets.readthedocs.io/en/latest/user_install.html and perhaps when I saw that jupyter command I thought something like *oh I'm running jupyterhub not jupyter, so I have to translate that in some way*. The suggestions I have to avoid others fall in the same trap are:

* reach out to the extension authors telling them to write in their docs that the jupyter command applies to jupyterhub too
* write somewhere (not sure where) in the jupyterhub docs that
  * you don't run jupyter as a server anymore, but
  * jupyter has a "configuration" mode (or whatever is called that `jupyter nbextension enable` thing) and you **do** run that in the same way, so *if you find docs from extensions (and anything else??) that are for jupyter and not specifically jupyterhub, you still run them as documented in this "configuration" mode*

Thanks again for giving helpful hints and for taking the lead on improving the docs.

**Comment by willingc** [comment]
@davidedelvento Thanks for the helpful feedback! Feel free to reach out on the Gitter channel if you hit any other roadblocks :sunny:

**Comment by minrk** [comment]
That's a good point. Something that could probably be clearer is that JupyterHub spawns regular single-user Jupyter notebook servers, so everything a user would do to configure their server (including extensions, etc.) applies in the same way to their server as launched by JupyterHub.

JupyterHub makes one very specific modification to the single-user server: it enables authentication via JupyterHub. Everything else should be the same.

**Comment by willingc** [comment]
Add to docs:

> JupyterHub spawns regular single-user Jupyter notebook servers, so everything a user would do to configure their server (including extensions, etc.) applies in the same way to their server as launched by JupyterHub.

JupyterHub makes one very specific modification to the single-user server: it enables authentication via JupyterHub.
